src/Evolution/DiscontinuousGalerkin/Filtering.hpp

```diff
@@ -0,0 +1,272 @@
+  ///
+  /// This option exists to temporarily disable the filter for debugging
+  /// purposes. For problems where filtering is not needed, the preferred
+  /// approach is to not compile the filter into the executable.
+  struct DisableForDebugging {
+    using type = bool;
+    static type default_value() noexcept { return false; }
+    static constexpr OptionString help = {"Disable the filter"};
+  };
+
+  using type = ExponentialFilter;
+  static std::string name() noexcept {
+    return "ExpFilter" + std::to_string(FilterIndex);
+  }
+
+  using options = tmpl::list<Alpha, HalfPower, DisableForDebugging>;
+  static constexpr OptionString help = {
+      "An exponential filter.\n"
+      // "The filter is applied in each direction to each tag."
+  };
+
+  ExponentialFilter() = default;
+
+  ExponentialFilter(double alpha, unsigned half_power,
+                    bool disable_for_debugging) noexcept;
+
+  // Action part of the class
+  template <typename DbTags, typename... InboxTags, typename ArrayIndex,
+            typename ActionList, typename ParallelComponent,
+            typename Metavariables>
+  static std::tuple<db::DataBox<DbTags>&&> apply(
+      db::DataBox<DbTags>& box,
+      const tuples::TaggedTuple<InboxTags...>& /*inboxes*/,
+      const Parallel::ConstGlobalCache<Metavariables>& cache,
+      const ArrayIndex& /*array_index*/, const ActionList /*meta*/,
+      const ParallelComponent* const /*meta*/) noexcept {
+    constexpr size_t volume_dim = Metavariables::system::volume_dim;
+    const ExponentialFilter& filter_helper =
+        Parallel::get<ExponentialFilter>(cache);
+    const Mesh<volume_dim> mesh = db::get<::Tags::Mesh<volume_dim>>(box);
+    DataVector temp(mesh.number_of_grid_points(), 0.0);
+    const Matrix empty{};
+    auto filter = make_array<volume_dim>(std::cref(empty));
+    for (size_t d = 0; d < volume_dim; d++) {
+      gsl::at(filter, d) = std::cref(filter_helper.filter_matrix(mesh, d));
+    }
+
+    db::mutate<TagsToFilter...>(
+        make_not_null(&box),
+        [&filter, &temp ](
+            const gsl::not_null<db::item_type<TagsToFilter>*>... filter_tensors,
+            const auto& local_mesh) noexcept {
+          const auto helper =
+              [&local_mesh, &filter, &temp ](const auto tensor) noexcept {
+            for (auto& component : *tensor) {
+              temp = 0.0;
+              apply_matrices(make_not_null(&temp), filter, component,
+                             local_mesh.extents());
+              component = temp;
+            }
+          };
+          EXPAND_PACK_LEFT_TO_RIGHT(helper(filter_tensors));
+        },
+        mesh);
+    return {std::move(box)};
+  }
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) noexcept;
+
+  bool operator==(const ExponentialFilter& rhs) const noexcept;
+
+ private:
+  template <size_t Dim>
+  const Matrix& filter_matrix(const Mesh<Dim>& mesh, size_t d) const noexcept {
+    switch (mesh.basis(d)) {
```

Comment:
I don't understand what you mean by "these". Could you clarify? The switch statements are used to translate from the runtime mesh values compile-time template parameters. We need the mesh to be runtime so that it is easier to use and because we want the ability to have different meshes on different elements and element interfaces.

Reply:
I see - I have misunderstood the mesh interface, and thought `Mesh`es were also templated on basis and quadrature.